File: python/src/application.py
import calendar
import datetime
import os
from typing import Iterator


class Application:
    """Generates working-report markdown templates for a username under
    `../../working-report/<username>/<year>/`, one file per day, week, or
    month."""

    class ValueError(Exception):
        pass

    class DigitLengthError(Exception):
        pass

    USERNAMES = ('hayat01sh1da',)
    with open(os.path.join('..', 'markdown', 'body_template.md')) as _f:
        BODY_TEMPLATE = _f.read()
    del _f
    FULL_UNITS = {'d': 'daily', 'w': 'weekly', 'm': 'monthly'}
    MONTHS_WITH_30_DAYS = ('April', 'June', 'September', 'November')

    @classmethod
    def run(cls, username: str = 'hayat01sh1da', unit: str = 'd',
            year: str | None = None) -> None:
        if year is None:
            year = datetime.date.today().strftime('%Y')
        instance = cls(username=username, unit=unit, year=year)
        instance.validate_username()
        instance.validate_unit()
        instance.validate_year()
        instance._run()

    def __init__(self, username: str = 'hayat01sh1da', unit: str = 'd',
                 year: str | None = None) -> None:
        self._username = username
        self._unit = unit
        self._year = year if year is not None else \
            datetime.date.today().strftime('%Y')

    def validate_username(self) -> None:
        if self._username not in self.USERNAMES:
            raise self.ValueError(
                f'{self._username} is NOT a permitted username.')

    def validate_unit(self) -> str:
        match self._unit:
            case 'd' | 'w' | 'm':
                return self._unit
            case _:
                raise self.ValueError('Provide d, w or m as a valid unit.')

    def validate_year(self) -> str:
        int(self._year)
        if len(self._year) > 4:
            raise self.DigitLengthError('Year must be 4 digits.')
        if int(self._year) < int(datetime.date.today().strftime('%Y')):
            raise self.ValueError(
                'Provide newer than or equal to the current year.')
        return self._year

    # private

    def _run(self) -> None:
        for i, month in enumerate(self._month_names(), start=1):
            index = f'{i:02}'
            directory = os.path.join(
                '..', '..', 'working-report', self._username,
                self._year, f'{index}_{month}',
            )
            self._process_month(month, i, index, directory)

    def _month_names(self) -> list[str]:
        return [calendar.month_name[i] for i in range(1, 13)]

    def _process_month(self, month: str, month_index: int,
                       index: str, directory: str) -> None:
        match self._unit:
            case 'd' | 'w':
                self._process_days(month, month_index, index, directory)
            case 'm':
                self._export_template(
                    directory=directory, index=index, month=month)

    def _process_days(self, month: str, month_index: int,
                      index: str, directory: str) -> None:
        for d in self._create_templates(month):
            if self._skip_day(month_index, d):
                continue
            day = f'{d:02}'
            self._export_template(
                directory=directory, index=index, day=day, month=month)

    def _skip_day(self, month_index: int, day: int) -> bool:
        if self._unit == 'd':
            return self._weekend(month_index, day)
        return not self._monday(month_index, day)

    def _full_unit(self) -> str:
        return self.FULL_UNITS.get(self._unit, '')

    def _monday(self, month: int, day: int) -> bool:
        return datetime.date(int(self._year), month, day).weekday() == 0

    def _weekend(self, month: int, day: int) -> bool:
        return datetime.date(int(self._year), month, day).weekday() >= 5

    def _leap_year(self) -> bool:
        y = int(self._year)
        return y % 400 == 0 or (y % 100 != 0 and y % 4 == 0)

    def _body(self, date: str) -> str:
        return self.BODY_TEMPLATE % date

    def _export_template(self, directory: str = '', index: str = '',
                         day: str = '', month: str = '') -> None:
        date_parts: list[str] = []
        if day:
            date_parts.append(f'{day} ')
        date_parts.append(f'{month} {self._year}')
        filename = os.path.join(
            directory,
            f'{self._year}{index}{day}_{self._full_unit()}'
            '_working_report.md',
        )
        os.makedirs(directory, exist_ok=True)
        if not os.path.exists(filename):
            with open(filename, 'w') as f:
                f.write(self._body(''.join(date_parts)))

    def _create_templates(self, month: str) -> Iterator[int]:
        for d in range(1, 32):
            if self._skip_day_of_month(month, d):
                continue
            yield d

    def _skip_day_of_month(self, month: str, day: int) -> bool:
        if month == 'February':
            return day > (29 if self._leap_year() else 28)
        if month in self.MONTHS_WITH_30_DAYS:
            return day > 30
        return False

File: ruby/src/application.rb
# frozen_string_literal: true
# rbs_inline: enabled

require 'date'

class Application
  class ValueError < StandardError; end
  class DigitLengthError < StandardError; end

  USERNAMES           = ['hayat01sh1da'].freeze
  BODY_TEMPLATE       = File.read(File.join(__dir__, 'body_template.md')).freeze
  FULL_UNITS          = { 'd' => 'daily', 'w' => 'weekly', 'm' => 'monthly' }.freeze
  MONTHS_WITH_30_DAYS = %w[April June September November].freeze

  # @rbs username: String
  # @rbs unit: String
  # @rbs year: String
  # @rbs return: void
  def self.run(username: 'hayat01sh1da', unit: 'd', year: Time.now.year.to_s)
    instance = new(username:, unit:, year:)
    instance.validate_username!
    instance.validate_unit!
    instance.validate_year!
    instance.run
  end

  # @rbs username: String
  # @rbs unit: String
  # @rbs year: String
  # @rbs return: void
  def initialize(username: 'hayat01sh1da', unit: 'd', year: Time.now.year.to_s)
    @username = username
    @unit     = unit
    @year     = year
  end

  # @rbs return: void
  def validate_username!
    raise ValueError, "#{username} is NOT a permitted username." unless USERNAMES.include?(username)
  end

  # @rbs return: String
  def validate_unit!
    case unit
    when 'd', 'w', 'm'
      unit
    else
      raise ValueError, 'Provide d, w or m as a valid unit.'
    end
  end

  # @rbs return: String
  def validate_year!
    Integer(year)
    if year.length > 4
      raise DigitLengthError, 'Year must be 4 digits.'
    elsif year.to_i < Time.now.year
      raise ValueError, 'Provide newer than or equal to the current year.'
    else
      year
    end
  end

  # @rbs return: void
  def run
    Date::MONTHNAMES.compact.each.with_index(1) do |month, i|
      index     = format('%02d', i)
      directory = File.join('..', '..', 'working-report', username, year, "#{index}_#{month}")
      process_month(month, i, index, directory)
    end
  end

  private

  attr_reader :username, :unit, :year

  # @rbs month: String
  # @rbs month_index: Integer
  # @rbs index: String
  # @rbs directory: String
  # @rbs return: void
  def process_month(month, month_index, index, directory)
    case unit
    when 'd', 'w' then process_days(month, month_index, index, directory)
    when 'm'      then export_template(directory:, index:, month:)
    end
  end

  # @rbs month: String
  # @rbs month_index: Integer
  # @rbs index: String
  # @rbs directory: String
  # @rbs return: void
  def process_days(month, month_index, index, directory)
    create_templates(month) do |d|
      next if skip_day?(month_index, d)

      day = format('%02d', d)
      export_template(directory:, index:, day:, month:)
    end
  end

  # @rbs month_index: Integer
  # @rbs day: Integer
  # @rbs return: bool
  def skip_day?(month_index, day)
    unit == 'd' ? weekend?(month_index, day) : !monday?(month_index, day)
  end

  # @rbs return: String
  def full_unit
    FULL_UNITS.fetch(unit, '')
  end

  # @rbs month: Integer
  # @rbs day: Integer
  # @rbs return: bool
  def monday?(month, day)
    Time.new(year, month, day).monday?
  end

  # @rbs month: Integer
  # @rbs day: Integer
  # @rbs return: bool
  def saturday?(month, day)
    Time.new(year, month, day).saturday?
  end

  # @rbs month: Integer
  # @rbs day: Integer
  # @rbs return: bool
  def sunday?(month, day)
    Time.new(year, month, day).sunday?
  end

  # @rbs month: Integer
  # @rbs day: Integer
  # @rbs return: bool
  def weekend?(month, day)
    date = Time.new(year, month, day)
    date.saturday? || date.sunday?
  end

  # @rbs return: bool
  def leap_year?
    (year.to_i % 400).zero? || (!!(year.to_i % 100).nonzero? && (year.to_i % 4).zero?)
  end

  # @rbs date: String
  # @rbs return: String
  def body(date)
    format(BODY_TEMPLATE, date)
  end

  # @rbs directory: String
  # @rbs index: String
  # @rbs day: String
  # @rbs month: String
  # @rbs array: Array[untyped]
  # @rbs return: void
  def export_template(directory: '', index: '', day: '', month: '', array: [])
    date     =  array
    date     << "#{day} " unless day.empty?
    date     << "#{month} #{year}"
    filename = File.join(directory, "#{year}#{index}#{day}_#{full_unit}_working_report.md")
    FileUtils.mkdir_p(directory)
    File.write(filename, body(date.join)) unless File.exist?(filename)
  end

  # @rbs month: String
  # @rbs return: void
  def create_templates(month, &)
    1.upto(31).each do |d|
      next if skip_day_of_month?(month, d)

      yield(d)
    end
  end

  # @rbs month: String
  # @rbs day: Integer
  # @rbs return: bool
  def skip_day_of_month?(month, day)
    case month
    when 'February'           then day > (leap_year? ? 29 : 28)
    when *MONTHS_WITH_30_DAYS then day > 30
    else false
    end
  end
end
